### ISSUE TITLE
create_dataset from data

### ISSUE THREAD

**Issue by clbarnes** [constantinpape/z5#30]
Draft implementation here:

https://github.com/constantinpape/z5/commit/14dd9a30edf7668b4b32b8e969c8b67ffe5acd44

My comments:

> It's convenient to do this. Only implemented in Base.create_dataset, as datasets shouldn't be created standalone anyway. Required giving defaults to some positional arguments. Added a h5py-like default dtype of 'float64'. If chunks is not given, the chunk size is the size of the passed-in array (so that it's contiguous like h5py).

@constantinpape 's thoughts

>This is mostly fine, however I am not really sure about two things:

>First the default datatype:

>I know having float64 as default is numpy / h5py convention, but personally I dislike it.

>In practice, the data we use is very rarely stored as float64 and having it as default
can oftentimes lead to not-so-easy to decipher errors occuring later, rather than
a meaningful error when creating the dataset.

> Same goes for the chunk-shape:

>Again, this is the h5py default, but for our use cases continuous storage is rarely the use-case.
So having this implicitly might result in sub-optimal performance due to one big chunk, without the user 
even noticing.

> I am open to discussion on these points though.

Discussion below

**Comment by clbarnes** [comment]
> I know having float64 as default is numpy / h5py convention, but personally I dislike it. In practice, the data we use is very rarely stored as float64 and having it as default
can oftentimes lead to not-so-easy to decipher errors occuring later, rather than
a meaningful error when creating the dataset.

I agree, but I think that h5py interoperation is really important and will really help buy-in with the format. It's a little bit out there already because the order of arguments is different, but at least having the same defaults moves them in the same direction.

> our use cases continuous storage is rarely the use-case. So having this implicitly might result in sub-optimal performance due to one big chunk, without the user
even noticing.

This is fair enough, but also true in HDF5 (although I guess HDF5 tends to target smaller data sets).

How about as a compromise, if we fallback to a default dtype or chunk size we raise a warning? Then we have A) the same tangible behaviour as h5py, B) some defaults for people who just want to throw some numbers onto a disk, and C) a pythonic "explicit is better than implicit" outcome where people know that their code is bad and they should feel bad?

If people are passing a vast dataset into the writer then they must have that dataset in memory already (or mapped to a single file) so it's probably not a totally infeasible size. We could do some further checks in case they're passing in, an h5py dataset which already has a chunksize defined.

**Comment by constantinpape** [comment]
Ok, I would be fine with the following:

*chunks:* make them optional, but instead of not chunking if they are not passed as argument,
fall back to a default chunk size (for the given number of dimensions).
Note that using contiguous storage in combination with compression could result in 
unwanted behavior, so I really want to avoid users ending up with this setting unless
they explicitly enforce it. 
The default chunks could be the same as in h5py (which I think is the equivalent for 64^3 for 3D data, etc.).
In this case we don't need to throw a warning, as this strategy is reasonable.

*dtype:* have `float64` as default dtype, but throw a warning when it is not passed as argument.

**Comment by clbarnes** [comment]
Sounds good to me.

I can't find the h5py implementation of autochunking, but given chunks can't be bigger than the dataset, how about

```python
DEFAULT_CHUNK_SIZE = 64

if chunks is None:
    chunks = tuple(min(s, DEFAULT_CHUNK_SIZE) for s in shape)
```

**Comment by alimanfoo** [comment]
FWIW the autochunking code in zarr is [here](https://github.com/zarr-developers/zarr/blob/master/zarr/util.py#L42). This was lifted straight from h5py (which I think originally got it from pytables) except that I have tweaked the defaults to make chunks bigger because in my experience automatic chunk sizes from h5py are way too small and give very sub-optimal performance.

**Comment by alimanfoo** [comment]
Also FWIW ``chunks=True`` is the default in Zarr, which means autochunking. In h5py the default is ``chunks=None`` which means no chunking and no compression. I find I hardly ever want ``chunks=None`` so ``chunks=True`` I think is a better default.

**Comment by constantinpape** [comment]
@clbarnes sorry for my inaccuracies, but I meant that the size (`product(chunk_shape)`) should be roughly `64^3`. Your suggestion would  result in very small chunks for low number of dimensions, but very large chunks for high dimensions.
Using the same code as @alimanfoo in zarr is probably a good idea (thanks for the hint).

I am considering to put this into the c++ part, I will let you know once I had a closer look.

@alimanfoo Yes, I agree no chunking and no compression is hardly ever wanted.

**Comment by clbarnes** [comment]
Fair points all round. I'll leave my branch open for the time being and leave it up to how you want to proceed.

https://github.com/clbarnes/z5/tree/dataset-from-data

**Comment by constantinpape** [comment]
I have implemented creation with data on the release1.0 branch.
I have also changed the API of `create_dataset` to be more consistent with `h5py`.
In addition, I have implemented `require_dataset`.
The only differences left to `h5py` (as far as I can see):
* If `chunks=None`, the dataset will get default chunks, the option `chunks=True` is invalid, because it is redundant with `chunks=None`.
* There is no default datatype. If creation happens without data, a datatype must be specified. Also, if creation happens with data and a datatype is given, it must be agree with the one of the data.
